share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,565 @@
+        WriteShareGroupStateResponseData responseData = WriteShareGroupStateResponse.toErrorResponseData(topicId, partitionId, error, message);
+        return new CoordinatorResult<>(Collections.emptyList(), responseData);
+    }
+
+    // Visible for testing
+    Integer getLeaderMapValue(SharePartitionKey key) {
+        return this.leaderEpochMap.get(key);
+    }
+
+    // Visible for testing
+    Integer getStateEpochMapValue(SharePartitionKey key) {
+        return this.stateEpochMap.get(key);
+    }
+
+    // Visible for testing
+    ShareGroupOffset getShareStateMapValue(SharePartitionKey key) {
+        return this.shareStateMap.get(key);
+    }
+
+    // Visible for testing
+    CoordinatorMetricsShard getMetricsShard() {
+        return metricsShard;
+    }
+
+    private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue newData) {
+        // snapshot epoch should be same as last share snapshot
+        // state epoch is not present
+        Set<PersisterOffsetsStateBatch> currentBatches = soFar.stateBatchAsSet();
+        long newStartOffset = newData.startOffset() == -1 ? soFar.startOffset() : newData.startOffset();
+        int newLeaderEpoch = newData.leaderEpoch() == -1 ? soFar.leaderEpoch() : newData.leaderEpoch();
+
+        return new ShareGroupOffset.Builder()
+            .setSnapshotEpoch(soFar.snapshotEpoch())
+            .setStateEpoch(soFar.stateEpoch())
+            .setStartOffset(newStartOffset)
+            .setLeaderEpoch(newLeaderEpoch)
+            .setStateBatches(combineStateBatches(currentBatches, newData.stateBatches().stream()
+                .map(PersisterOffsetsStateBatch::from)
+                .collect(Collectors.toCollection(LinkedHashSet::new)), newStartOffset))
+            .build();
+    }
+
+    /**
+     * Util method which takes in 2 collections containing {@link PersisterOffsetsStateBatch}
+     * and the startOffset.
+     * It removes all batches from the 1st collection which have the same first and last offset
+     * as the batches in 2nd collection. It then creates a final list of batches which contains the
+     * former result and all the batches in the 2nd collection. In set notation (A - B) U B (we prefer batches in B
+     * which have same first and last offset in A).
+     * Finally, it removes any batches where the lastOffset < startOffset, if the startOffset > -1.
+     * @param currentBatch - collection containing current soft state of batches
+     * @param newBatch - collection containing batches in incoming request
+     * @param startOffset - startOffset to consider when removing old batches.
+     * @return List containing combined batches
+     */
+    private static List<PersisterOffsetsStateBatch> combineStateBatches(
+        Collection<PersisterOffsetsStateBatch> currentBatch,
+        Collection<PersisterOffsetsStateBatch> newBatch,
+        long startOffset
+    ) {
```

Comment:
I think we can tighten up this signature. We know what types are expected, so we can specify them here. That will make the code below a bit more understandable I think.

I'd also like to see some unit tests for this method. It seems like a good candidate.